docs/OBJECT-LEVEL-ENCRYPTION.md

```diff
@@ -0,0 +1,119 @@
+
+Downloader passes the cipher context and controller certificate subscription contexts and cipher block from the datastore config to the cipher package for the decryption.
+
+The datastore credentials are only kept in memory in the downloader.
+
+### Network Interface Manager
+
+Nim needs to decrypt the WiFi credentials available in device port configuration for setting up the wlan connectivity. Nim calls the [device network](../pkg/pillar/devicenetwork) package for parsing the device port configuration.
+
+Nim passes device port configuration and subscription contexts of controller certificates and cipher contexts to the device network package. And, device network package passes received subscription contexts and cipher block from the WiFi configuration to the cipher package for the decryption.
+
+Decryption of WiFi credentials is very important if the device has only wlan connectivity, and depends on the key information being published persistently so that the ECDH calculation and decryption can be performed on reboot before the device can communication with the controller.
+
+The WiFi credentials are only kept in memory in nim, however in order to hand them to the wpa_supplicant process they are stored in /run which is a filesystem backed by memory.
+
+### Domain Manager
+
+Domain manager needs to decrypt the sensitive user data from the app instance configuration used for cloud-init and for container environment settings. Depending on the application instance such information can include local and remove keys and credentials.
+
+Domain manager passes the cipher block from the app configuration and subscription contexts of controller certificates and cipher context to the cipher package for the decryption.
+
+The user data are only kept in memory in the domainmgr, but for cloud-init it is used to create a CDROM image which is passed to the application. This CDROM image lives in /var/run/domainmgr/cloudinit which is a filesystem backed by memory.
+
+## Refreshing of the controller certificates in EVE
+
+Zedagent fetches the controller certificates at boot. So, we need a mechanism of refreshing the controller certificates in EVE if the controller certificates get updated when the device is already running.
+
+There are two ways for getting updated certificates:
+
+1. We can include hashes of the certificates on the top level of the device configuration itself. With the help of it, we can parse the certificates from the the device configuration in zedagent and will fetch the certificates if changed.
+2. If we don’t want to add the hashes of certificates in the the device configuration then we can add a decryption retry mechanism in the cipher package. So if decryption fails because of a certificate not found error, we will identify that the certificate required for the decryption is not available in the device and will fetch the controller certificates and trigger the decryption retry mechanism.
```

Comment:
This one is pretty clear, but it makes it sound like, "well, we _really_ don't want this, one but if we _have_ to, ok, this will work too." I don't know if that is the intent or not, but if not, then let's be assertive about it:

"Use retries. If decryption fails because of a certificate not found error, fetch latest controller certificates and retry."

Reply:
Part of TBD